scripts/core/EXP-COMMAND-BUFFER.rst

```diff
@@ -138,14 +140,45 @@ were obtained from.
     ${x}_exp_command_buffer_sync_point_t syncPoint;
 
     ${x}CommandBufferAppendUSMMemcpyExp(hCommandBuffer, pDst, pSrc, size, 0,
-                                        nullptr, &syncPoint);
+                                        nullptr, 0, nullptr, &syncPoint, nullptr,
+                                        nullptr);
 
     // Append a kernel launch with syncPoint as a dependency, ignore returned
     // sync-point
     ${x}CommandBufferAppendKernelLaunchExp(hCommandBuffer, hKernel, workDim,
                                            pGlobalWorkOffset, pGlobalWorkSize,
                                            pLocalWorkSize, 1, &syncPoint,
-                                           nullptr, nullptr);
+                                           nullptr, 0, nullptr, nullptr,
+                                           nullptr);
+
+Command Synchronization With Events
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+When appending commands to a command-buffer an optional ``phEventWaitList``
+input parameter is available for passing a list of ${x}_event_handle_t objects
+the command should wait on. As well as an optional ``phEvent`` output parameter
+to get a ${x}_event_handle_t object that will be signaled the next time the
+command executes. This returned signal event is only valid for last execution
+of the command, and prior to the first execution of the command-buffer it
+represents an empty submission that is considered complete. It is the users
+responsibility to release the returned ``phEvent`` with ${x}EventRelease.
+
+These parameters allow commands in a command-buffer to depend on the completion
+of UR commands submitted to a queue which are external to a command-buffer. As
+well as allowing such commands to depend on individual command-buffer commands
+by themselves waiting on the event returned by commands.
+
+It is also possible for commands in different command-buffer objects to
+synchronize using this mechanism. However, this is only possible in one
+direction where the signal event of one command-buffer command is used as a
+wait event of another command-buffer command. Bi-directional sync between
+individual commands in the two command-buffers is not possible due to the
+completion state of the command events only being reset when a command-buffer
+is executed.
```

Comment:
Yeah this wording isn't very explanatory and needs refined. This came up during doing the CUDA backend, and it's the following scenario i'm struggling with:

1. cmdbuf 1 -Append fill to ptr A
2. cmdbuf 1 - Append fill to ptr B waiting on previous command using sync point, returning an output event (E1).
3. cmdbuf 2 - kernel using ptr A & B as inputs writing ptr C, waiting on E1 event from step 2 and outputting new event (E2).
4. cmdbuf 1 - Read of ptrC back to host waiting on E2


Then finalization would give two linear CUDA-Graphs:
G1: Fill->Fill->Signal E1->Wait E2->Read
G2: Wait E1->Kernel->Signal->E2

E1 and E2 are empty after finalize and default to complete.

If a user enqueues G2 then G1, the first wait node could finish straight aware are E1 is complete and G2 could complete before G1 starts.
If a user enqueues G1 then G2, then the it could also finish before G2 starts as E2 is complete.

I'm struggling to work out how to get the desired semantics of two concurrently executing graphs which wait on each other with a CUDA backend.

Reply:
I've pushed a spec update to clarify why this is
```quote
Bi-directional sync between individual commands in two separate command-buffers
is however not guaranteed to behave correctly. This is due to the completion
state of the command events only being reset when a command-buffer is executed.
It is therefore possible for the first command-buffer enqueued to execute the
wait node that needs to be reset by the enqueue of the second command-buffer,
before the code path returns to user code for the user to enqueue the second
command-buffer.
```